mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py

```diff
@@ -0,0 +1,209 @@
+        if pre_topk > 0:
+            if with_score_factors:
+                nms_pre_score = nms_pre_score * score_factors
+            if backend == Backend.TENSORRT:
+                priors = pad_with_value(priors, 1, pre_topk)
+                bbox_pred = pad_with_value(bbox_pred, 1, pre_topk)
+                scores = pad_with_value(scores, 1, pre_topk, 0.)
+                nms_pre_score = pad_with_value(nms_pre_score, 1, pre_topk, 0.)
+                if with_score_factors:
+                    score_factors = pad_with_value(score_factors, 1, pre_topk,
+                                                   0.)
+
+            # Get maximum scores for foreground classes.
+            if self.use_sigmoid_cls:
+                max_scores, _ = nms_pre_score.max(-1)
+            else:
+                max_scores, _ = nms_pre_score[..., :-1].max(-1)
+            _, topk_inds = max_scores.topk(pre_topk)
+            batch_inds = torch.arange(
+                batch_size,
+                device=bbox_pred.device).view(-1, 1).expand_as(topk_inds)
+            priors = priors[batch_inds, topk_inds, :]
+            bbox_pred = bbox_pred[batch_inds, topk_inds, :]
+            scores = scores[batch_inds, topk_inds, :]
+            if with_score_factors:
+                score_factors = score_factors[batch_inds, topk_inds, :]
+
+        mlvl_valid_bboxes.append(bbox_pred)
+        mlvl_valid_scores.append(scores)
+        priors = self.anchor_center(priors)  # xyxy2xy by richard.
+        mlvl_valid_priors.append(priors)
+        if with_score_factors:
+            mlvl_score_factors.append(score_factors)
+
+    batch_mlvl_bboxes_pred = torch.cat(mlvl_valid_bboxes, dim=1)
+    batch_scores = torch.cat(mlvl_valid_scores, dim=1)
+    batch_priors = torch.cat(mlvl_valid_priors, dim=1)
+    batch_bboxes = self.bbox_coder.decode(
+        batch_priors, batch_mlvl_bboxes_pred, max_shape=img_shape)
+    if with_score_factors:
+        batch_score_factors = torch.cat(mlvl_score_factors, dim=1)
+
+    if not self.use_sigmoid_cls:
+        batch_scores = batch_scores[..., :self.num_classes]
+
+    if with_score_factors:
+        batch_scores = batch_scores * batch_score_factors
+    if not with_nms:
+        return batch_bboxes, batch_scores
+    post_params = get_post_processing_params(deploy_cfg)
+    max_output_boxes_per_class = post_params.max_output_boxes_per_class
+    iou_threshold = cfg.nms.get('iou_threshold', post_params.iou_threshold)
+    score_threshold = cfg.get('score_thr', post_params.score_threshold)
+    pre_top_k = post_params.pre_top_k
+    keep_top_k = cfg.get('max_per_img', post_params.keep_top_k)
+    return multiclass_nms(
+        batch_bboxes,
+        batch_scores,
+        max_output_boxes_per_class,
+        iou_threshold=iou_threshold,
+        score_threshold=score_threshold,
+        pre_top_k=pre_top_k,
+        keep_top_k=keep_top_k)
+
+
+def integral(x, reg_max=16, project=None):
```

Comment:
It would be better to write a integral with batch support. for example:

```python

    def batched_integral(self, intergral, x):
        batch_size = x.size(0)
        x = F.softmax(x.reshape(batch_size, -1, intergral.reg_max + 1), dim=2)
        x = F.linear(x,
                     intergral.project.type_as(x).unsqueeze(0)).reshape(
                         batch_size, -1, 4)
        return x
```

Reply:
Thank you so much. I modified and tested the code as you suggested and it works fine. But the shape of `bbox_pred` obtained by `forward_single` is B, C, H, W. I tried to execute `bbox_pred.permute(0,2,3,1)` in `forward_single` and `get_bbox` respectively, and test the speed, the former is faster.

Reply:
Your suggestion helped me a lot, however, I'm not familiar with mmdeploy's uint test, so I'm sorry I can't give more help.  
I am able to provide the command that runs when I test.  
> - deploy:  
```python tools/deploy.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py ../MMDetection/checkpoints/gfl_r50_fpn_1x_coco_20200629_121244-25944287.pth ../MMDetection/demo/demo.jpg --work-dir test_work/test_GFL/ --show --device cuda:0```    
> - test:  
```python tools/test.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py --model test_work/test_GFL/end2end.engine --speed-test --device cpu```

Reply:
> Thank you so much. I modified and tested the code as you suggested and it works fine. But the shape of `bbox_pred` obtained by `forward_single` is B, C, H, W. I tried to execute `bbox_pred.permute(0,2,3,1)` in `forward_single` and `get_bbox` respectively, and test the speed, the former is faster.

Thanks for the test. How faster if we use nonbatched integral? I think it is worth supporting batch inference with a little performance.